rfd/0125-dynamic-auto-discovery-config.md

```diff
@@ -0,0 +1,406 @@
+
+Running `discovery_service`s in high availability (multiple instances and same config) should yeld the same result as before.
+
+### New `discovery_service` property: `selector`
+The configuration of the `discovery_service` has a new field: `selector`.
+
+This field works as a label matcher for `DiscoveryConfig` resources.
+Standard label matcher rules apply here (eg, Role's `node_labels` or Database Service's `resources.labels`).
+
+For each matching `DiscoveryConfig`, the list of matchers is appended to the `discovery_service` matchers.
+
+Example:
+```yaml
+discovery_service:
+  enabled: "yes"
+  discovery_group: my_group
+  selector:
+    cloud: aws
+    env: [qa, prod]
+  aws:
+  - types: ["ec2"]
+    regions: ["us-east-1","us-west-1"]
+    # ...
+  azure:
+  - types: ["aks"]
+    regions: ["eastus", "westus"]
+    # ...
+  gcp:
+  - types: ["gke"]
+    # ...
+```
+
+### Hot-reloading `DiscoveryConfig` matchers
+When the `discovery_service` starts, it loads all the static matchers plus the dynamic ones (using `DiscoveryConfig` resources).
+
+When the user changes (either creating, deleting or updating) a `DiscoveryConfig`, the `discovery_service` must update its discovery matchers.
+
+To achieve this, `discovery_service` must create a Watcher over the `DiscoveryConfig` resource and when a new change occurs update its internal list of matchers.
+
+```go
+	watcher, _ := clt.NewWatcher(ctx, types.Watch{
+		Kinds: []types.WatchKind{{Kind: types.KindDiscoveryConfig}},
+	})
+
+	defer watcher.Close()
+
+	for {
+		select {
+		case event := <-watcher.Events():
+			// ...
+			discoveryConfig := event.Resource
+			switch event.Type {
+			case types.OpDelete:
+				discoveryWatchers.RemoveByDiscoveryConfigName(discoveryConfig.GetName())
+			case types.OpPut:
+				discoveryWatchers.UpsertDiscoveryConfig(discoveryConfig)
+			}
+
+			return nil
+
+		case <-watcher.Done():
+			if err := watcher.Error(); err != nil {
+				return trace.Wrap(err)
+			}
+			return nil
+		case <-ctx.Done():
+			return nil
+		}
+	}
+```
+
+### Proto Specification for `DiscoveryConfig`
+`DiscoveryConfig` will have the following protobuf specification
+
+```proto
+// DiscoveryConfigV1 represents a DiscoveryConfig to be used by discovery_services.
+message DiscoveryConfigV1 {
+  // Header is the resource header.
+  ResourceHeader Header = 1;
+
+  // Spec is an DiscoveryConfig specification.
+  DiscoveryConfigSpecV1 Spec = 2 ;
+}
+
+// DiscoveryConfigSpecV1 contains properties required to create matchers to be used by discovery_service.
+// Those matchers are used by discovery_service to watch for cloud resources and create them in Teleport.
+message DiscoveryConfigSpecV1 {
+  // AWS is a list of AWSMatchers.
+  repeated AWSMatcher aws = 1;
+
+  // Azure is a list of AzureMatchers.
+  repeated AzureMatcher azure = 2;
+
+  // GCP is a list of GCPMatchers.
+  repeated GCPMatcher gcp = 3; 
+}
+
+// AWSMatcher matches AWS EC2 instances and AWS Databases
+message AWSMatcher {
```

Comment:
Just noting that we should try and reuse all the matcher proto structs we already have defined instead of adding new ones.